File: totp_algo.py
import hashlib
import hmac
import math
import time

def getSlicedTimeStamp(interval=30):
    now_in_seconds = math.floor(time.time())
    return math.floor(now_in_seconds / interval)


def generateTimeBasedHash(key, message):
    hash = hmac.new(
        bytes(key, encoding="utf-8"),
        bytes(message, encoding='utf-8'),
        hashlib.sha256,
    )
    return hash.hexdigest()


def dynamicTruncation(raw_key, length=6):
    offset = int(raw_key[0], base=16)
    chosen32bits = raw_key[offset: offset+8]
    otpstr = str(int(chosen32bits, base=16))
    return otpstr[-length:]


def generateTOTP(key, interval):
    timestamp = getSlicedTimeStamp(interval)
    raw_key = generateTimeBasedHash(key, str(timestamp))
    return dynamicTruncation(raw_key)
    


shared_key = 'e8fb1a2faf331bfffe8670ca20447fae'
interval = 30
totp = generateTOTP(shared_key, interval)
print(totp)

File: totp_algo.js
const crypto = require('crypto')

function getSlicedTimeStamp(interval=30){
    const timeInSeconds = Date.now()/1000
    return parseInt(String(timeInSeconds / interval)).toString()
}

function generateTimeBasedHash(key, message){
    const hash = crypto.createHmac("SHA256", key).update(message).digest()
    return hash.toString('hex')
}

function dynamicTruncation(hexDigest, length = 6){
    const offset = parseInt( hexDigest[0], 16)
    const chosen32bits = hexDigest.slice(offset, offset+8)
    const otpString = BigInt(parseInt(chosen32bits, 16)).toString(10)
    const n = otpString.length
    return otpString.slice(n-length, n)
}

function generateTOTP(key, intervalInSeconds){
    const timestamp = getSlicedTimeStamp(intervalInSeconds)
    const hexDigest = generateTimeBasedHash(key, timestamp)
    return dynamicTruncation(hexDigest)
}

async function main(){
    const intervalInSeconds = 30
    const shared_key = 'e8fb1a2faf331bfffe8670ca20447fae'
    const totp = generateTOTP(shared_key, intervalInSeconds)
    console.log(totp);
}

main().then(()=>{console.log('completed!');}).catch((err)=>{console.log('ERROR: ', err);})